2015-10-31-dekorator.md

```diff
@@ -0,0 +1,185 @@
+dekorator, który umożliwia nam rozszerzanie funkcjonalności obiektu w czasie pracy aplikacji. 
+{% endblock %}
+
+{% block dummy_implementation_php %}
+```language-php
+interface SomeFancyInterface
+{
+    public function getInfo();
+}
+
+class BasicObject implements SomeFancyInterface
+{
+    public function getInfo()
+    {
+        return "I'm basic object";
+    }
+}
+
+class Decorator implements SomeFancyInterface
+{
+    private $object;
+
+    public function __construct($object)
+    {
+        $this->object = $object;
+    }
+
+    public function getInfo()
+    {
+        return $this->object->getInfo() . " decorated!";
+    }
+}
+
+$object = new BasicObject();
+$object = new Decorator($object);
+
+echo $object->getInfo() . PHP_EOL; // I'm basic object decorated!
+```
+{% endblock %}
+
+{% block dummy_implementation_ruby %}
+```language-ruby
+
+```
+{% endblock %}
+
+{% block dummy_implementation_java %}
+```language-java
+
+```
+{% endblock %}
+
+{% block dummy_implementation_c_sharp %}
+```language-csharp
+
+```
+{% endblock %}
+
+{% block when_to_use %}
+Dekorator jest niezwykle użyteczny w momencie kiedy naturalnym zachowaniem pewnego obiektu jest jego podatność na 
+rozszerzenia. Można wyobrazić sobie wiele przypadków użycia, jak na przykład proces zakupu samochodu (dodwanie nowych
+opcji, jak nawigacja, przyciemniane szyby), wspomniana wyżej pizzeria, polisa ubezpieczeniowa itp. Zaprojektujmy zatem
+banalny model dla sprzedaży telewizji kablowej, skupmy się na ofercie.
+{% endblock %}
+
+{% block example_specification %}
+- pkt 1. Mamy podstawowaą ofertę
+- pkt 2. Klient może do oferty dodać opcje: pakiet sportowy, edukacyjny i filmowy
+- pkt 3. Zbudujmy finalną ofertę za pomocą dekoratorów
+{% endblock %}
+
+{% block example_code_php %}
+```language-php
+<?php
+
+interface Offer
+{
+    public function getPrice();
+}
+
+final class SmallOffer implements Offer
+{
+
+    public function getPrice()
+    {
+        return 40;
+    }
+}
+
+final class SportDecorator implements Offer
+{
+    private $offer;
+
+    public function __construct($offer)
+    {
+        $this->offer = $offer;
+    }
+
+    public function getPrice()
+    {
+        return $this->offer->getPrice() + 20;
+    }
+}
+
+final class EducationDecorator implements Offer
+{
+    private $offer;
+
+    public function __construct($offer)
+    {
+        $this->offer = $offer;
+    }
+
+    public function getPrice()
+    {
+        return $this->offer->getPrice() + 10;
+    }
+}
+
+final class CinemaDecorator implements Offer
+{
+    private $offer;
+
+    public function __construct($offer)
+    {
+        $this->offer = $offer;
+    }
+
+    public function getPrice()
+    {
+        return $this->offer->getPrice() + 30;
+    }
+}
+
+$offer = new SmallOffer();
+$offer = new SportDecorator($offer);
+$offer = new CinemaDecorator($offer);
+$offer = new EducationDecorator($offer);
```

Comment:
Oki, ale właśnie Twój przykład generuje kilka ofert na jednym dekoratorze, to już może lepiej:

```
$baseOffer = new BaseOffer();
$offerWithSport = new SportDecorator($baseOffer);
$offerWithSportAndEducation = new CinemaDecorator(new SportDecorator($baseOffer));
$fullOffer = new EducationDecorator(new CinemaDecorator(new SportDecorator($baseOffer)));

echo $baseOffer->getPrice() . PHP_EOL; // 40
echo $offerWithSport->getPrice() . PHP_EOL; // 60
echo $offerWithSportAndEducation->getPrice() . PHP_EOL; // 90
echo $fullOffer->getPrice() . PHP_EOL; // 100
```